src/main/java/picard/sam/markduplicates/UmiAwareDuplicateSetIterator.java

```diff
@@ -0,0 +1,225 @@
+    int editDistanceToJoin;
+    boolean addInferredUmi;
+
+    public UmiAwareDuplicateSetIterator(final DuplicateSetIterator wrappedIterator, final int editDistanceToJoin, boolean addInferredUmi) {
+        this.wrappedIterator = wrappedIterator;
+        this.editDistanceToJoin = editDistanceToJoin;
+        this.addInferredUmi = addInferredUmi;
+        nextSetsIterator = Collections.emptyIterator();
+    }
+
+    @Override
+    public void close() {
+        wrappedIterator.close();
+    }
+
+    @Override
+    public boolean hasNext() {
+        return nextSetsIterator.hasNext() || wrappedIterator.hasNext();
+    }
+
+    @Override
+    public DuplicateSet next() {
+        if(!nextSetsIterator.hasNext())
+            process(wrappedIterator.next());
+
+        return nextSetsIterator.next();
+    }
+
+    // Takes a duplicate set and breaks it up into possible smaller sets according to the UMI,
+    // and updates nextSetsIterator to be an iterator on that set of DuplicateSets.
+    private void process(final DuplicateSet set) {
+
+        List<SAMRecord> records = set.getRecords();
+
+        groups = new ArrayList<>();
+        adjacencyList = new ArrayList<>();
+
+        // If any records don't have the RX attribute... don't do anything special
+        for(int i = 0;i < records.size();i++) {
+            if(records.get(i).getAttribute("RX") == null) {
+                nextSetsIterator = Collections.singleton(set).iterator();
+                return;
+            } else {
+            }
+        }
+
+        // Sort records by RX tag
+        Collections.sort(records, new Comparator<SAMRecord>() {
+            @Override
+            public int compare(final SAMRecord lhs, final SAMRecord rhs) {
+                if(lhs == null || rhs == null) {
+                    return 0;
+                } else if(lhs.getAttribute("RX") == null || rhs.getAttribute("RX") == null) {
+                    return 0;
+                }
+                else {
+                    return ((String) lhs.getAttribute("RX")).compareTo((String) rhs.getAttribute("RX"));
+                }
+            }
+        });
+
+        int n = records.size();
+
+        // Locate records that have identical UMI sequences
+        List<String> uniqueObservedUMIs = new ArrayList<String>();
+
+        uniqueObservedUMIs.add((String) records.get(0).getAttribute("RX"));
+        for(int i = 1;i < n;i++) {
+            // If the records differ (after sorting) we have a new duplicate set.
+            if(!records.get(i).getAttribute("RX").equals(records.get(i-1).getAttribute("RX"))) {
+                uniqueObservedUMIs.add((String) records.get(i).getAttribute("RX"));
+            }
+        }
+
+        // Construct Adjacency List of UMIs that are close
+        for(int i = 0;i < uniqueObservedUMIs.size();i++) {
+            adjacencyList.add(i, new ArrayList<Integer>());
+            groups.add(i, 0);
+            for(int j = 0;j < uniqueObservedUMIs.size();j++) {
```

Comment:
no, I guess it's fine. I didn't realize that there are two lists here...this is actually a little concerning since it relies on a connection between the two lists...and there's no check that the lists are properly connected. I'd recommend using a map instead of a list in one of the containers to alleviate this.